### PR TITLE
Cmake static or shared

### PR DESCRIPTION
**Pull request by autoantwort**
Same as #951 but merged with master

### CODE REVIEW

**Comment by prince-chrismc** [comment]
This looks really good!

**Comment by autoantwort** [comment]
@prince-chrismc Can you approve the ci again? :)

**Comment by prince-chrismc** [comment]
I don't have permission i just would benefit from this PR ❤️

**Comment by michael-grunder** [comment]
Thanks for rebasing!

This looks good to me but I'll cc @bjosv and @charsyam (our resident CMake experts :smile:) in case anything catches their eye.

File: examples/CMakeLists.txt [hunk 1]
```diff
@@ -16,46 +16,46 @@ FIND_PATH(LIBEV ev.h
...
     FIND_LIBRARY(CF CoreFoundation)
     ADD_EXECUTABLE(example-macosx example-macosx.c)
-    TARGET_LINK_LIBRARIES(example-macosx ${HIREDIS_DEFAULT_LIBRARY} ${CF})
+    TARGET_LINK_LIBRARIES(example-macosx hiredis ${CF})
 ENDIF()
 
 IF (ENABLE_SSL)
     ADD_EXECUTABLE(example-ssl example-ssl.c)
-    TARGET_LINK_LIBRARIES(example-ssl ${HIREDIS_DEFAULT_LIBRARY} ${HIREDIS_SSL_DEFAULT_LIBRARY})
+    TARGET_LINK_LIBRARIES(example-ssl hiredis ${HIREDIS_SSL_DEFAULT_LIBRARY})
```
Comment (bjosv) [inline examples/CMakeLists.txt:54]:
```suggestion
    TARGET_LINK_LIBRARIES(example-ssl hiredis hiredis_ssl)
```

**Review by bjosv — COMMENTED** [review]
Nice, this PR makes it look a lot cleaner.

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -82,12 +82,14 @@ jobs:
     steps:
       - uses: microsoft/setup-msbuild@v1.0.2
       - uses: actions/checkout@v3
-      - name: Run CMake
+      - name: Run CMake (shared lib)
+        run: cmake -Wno-dev CmakeLists.txt -DBUILD_SHARED_LIBS=ON
```
Comment (bjosv) [inline .github/workflows/test.yml:86]:
```suggestion
        run: cmake -Wno-dev CMakeLists.txt
``
The default BUILD_SHARED_LIBS is already ON?

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -82,12 +82,14 @@ jobs:
...
+      - name: Build hiredis (shared lib)
+        run: MSBuild hiredis.vcxproj /p:Configuration=Debug
+      - name: Run CMake (static lib)
         run: cmake -Wno-dev CmakeLists.txt
```
Comment (bjosv) [inline .github/workflows/test.yml:90]:
```suggestion
        run: cmake -Wno-dev CMakeLists.txt -DBUILD_SHARED_LIBS=OFF
```

**Review by bjosv — COMMENTED** [review]
Another thing, the option `BUILD_SHARED_LIBS` seems already to be defaulted to ON.
The CI needs to be changed in that case.
(Also found a legacy nit regarding filename)

**Review by chayim — APPROVED** [review]

**Comment by michael-grunder** [comment]
@autoantwort @chayim Assuming there are no objections, I'll probably squash and merge this PR since there are so many commits.

**Review by bjosv — APPROVED** [review]
LGTM.

File: CMakeLists.txt [hunk 4]
```diff
@@ -45,50 +43,25 @@ IF(WIN32)
     ADD_DEFINITIONS(-D_CRT_SECURE_NO_WARNINGS -DWIN32_LEAN_AND_MEAN)
 ENDIF()
 
-ADD_LIBRARY(hiredis_static STATIC ${hiredis_sources})
-ADD_LIBRARY(hiredis::hiredis_static ALIAS hiredis_static)
```
[the remark this answers is not in the document]
Comment (autoantwort) [reply CMakeLists.txt:49]: Yeah that is true. I could add the following (behind an option) 
```cmake
set_target_properties(hiredis PROPERTIES EXPORT_NAME hiredis_static)
```
then you have your `hiredis::hiredis_static` target back.

[the remark this answers is not in the document]
Comment (dmitrmax) [reply CMakeLists.txt:49]: Ok. Go ahead please,

[the remark this answers is not in the document]
Comment (dmitrmax) [reply CMakeLists.txt:49]: JFYI, there is a brilliant article for libraries' authors about exporting shared/static configurations in CMake. 

https://alexreinking.com/blog/building-a-dual-shared-and-static-library-with-cmake.html

[the remark this answers is not in the document]
Comment (michael-grunder) [reply CMakeLists.txt:49]: No rush @autoantwort but if you add the final change to allow building both I'll get this merged.

[the remark this answers is not in the document]
Comment (autoantwort) [reply CMakeLists.txt:49]: > https://alexreinking.com/blog/building-a-dual-shared-and-static-library-with-cmake.html

Nice read. 

@michael-grunder Done. Thank you for the ping, forgot it.

**Comment by michael-grunder** [comment]
Merged, thanks!

**Pull request merged**